File: web/src/export.ts
import { formatUSDC } from "./format";
import type { Report } from "./types";

const header = [
  "record_type",
  "reference",
  "status",
  "direction",
  "expected_usdc",
  "settled_usdc",
  "difference_usdc",
  "counterparty",
  "transaction_hash",
  "memo_id",
  "memo_text",
  "fee_usdc",
  "confidence",
  "reason",
];

export function reportToCSV(report: Report): string {
  const rows: string[][] = [header];
  for (const match of report.matches) {
    if (match.payments.length === 0) {
      rows.push([
        "expected",
        match.expected.reference,
        match.status,
        match.expected.direction,
        formatUSDC(match.expected.amountMicros),
        "0.00",
        formatUSDC(match.differenceMicros),
        match.expected.counterparty,
        "",
        match.expected.memoId ?? "",
        "",
        "0.00",
        match.confidence,
        match.reason,
      ]);
      continue;
    }
    for (const payment of match.payments) {
      rows.push([
        "matched",
        match.expected.reference,
        match.status,
        payment.direction,
        formatUSDC(match.expected.amountMicros),
        formatUSDC(payment.amountMicros),
        formatUSDC(match.differenceMicros),
        payment.direction === "inbound" ? payment.from : payment.to,
        payment.transactionHash,
        payment.memoId ?? "",
        payment.memoText ?? "",
        formatUSDC(payment.feeMicros),
        match.confidence,
        match.reason,
      ]);
    }
  }
  for (const payment of report.unmatchedPayments) {
    rows.push([
      "unmatched",
      "",
      "unmatched",
      payment.direction,
      "0.00",
      formatUSDC(payment.amountMicros),
      formatUSDC(payment.amountMicros),
      payment.direction === "inbound" ? payment.from : payment.to,
      payment.transactionHash,
      payment.memoId ?? "",
      payment.memoText ?? "",
      formatUSDC(payment.feeMicros),
      "none",
      "no expected record matched this onchain payment",
    ]);
  }
  return rows.map((row) => row.map(csvCell).join(",")).join("\r\n") + "\r\n";
}

function csvCell(value: string): string {
  const safe = /^[=+\-@]/.test(value.trimStart()) ? `'${value}` : value;
  return `"${safe.replaceAll('"', '""')}"`;
}

File: internal/reconcile/export.go
package reconcile

import (
	"encoding/csv"
	"encoding/json"
	"fmt"
	"io"
	"strings"
)

func WriteReportJSON(writer io.Writer, report Report) error {
	encoder := json.NewEncoder(writer)
	encoder.SetIndent("", "  ")
	return encoder.Encode(report)
}

func WriteReportCSV(writer io.Writer, report Report) error {
	w := csv.NewWriter(writer)
	defer w.Flush()
	if err := w.Write([]string{
		"record_type", "reference", "status", "direction", "expected_usdc", "paid_usdc",
		"difference_usdc", "counterparty", "transaction_hash", "memo_id", "memo_text",
		"fee_usdc", "confidence", "reason",
	}); err != nil {
		return err
	}
	for _, match := range report.Matches {
		if match.Expected == nil {
			continue
		}
		if len(match.Payments) == 0 {
			if err := w.Write(safeCSVRow([]string{
				"expected", match.Expected.Reference, string(match.Status), string(match.Expected.Direction),
				FormatUSDC(match.Expected.AmountMicros), "0.000000", FormatUSDC(match.DifferenceMicros),
				match.Expected.Counterparty, "", match.Expected.MemoID, "", "0.000000", match.Confidence, match.Reason,
			})); err != nil {
				return err
			}
			continue
		}
		for _, payment := range match.Payments {
			if err := w.Write(safeCSVRow([]string{
				"matched", match.Expected.Reference, string(match.Status), string(payment.Direction),
				FormatUSDC(match.Expected.AmountMicros), FormatUSDC(payment.AmountMicros), FormatUSDC(match.DifferenceMicros),
				paymentCounterparty(payment), payment.TransactionHash, payment.MemoID, payment.MemoText,
				FormatUSDC(payment.FeeMicros), match.Confidence, match.Reason,
			})); err != nil {
				return err
			}
		}
	}
	for _, payment := range report.UnmatchedPayments {
		if err := w.Write(safeCSVRow([]string{
			"unmatched", "", string(StatusUnmatched), string(payment.Direction), "0.000000",
			FormatUSDC(payment.AmountMicros), FormatUSDC(payment.AmountMicros), paymentCounterparty(payment),
			payment.TransactionHash, payment.MemoID, payment.MemoText, FormatUSDC(payment.FeeMicros),
			"none", "no expected record matched this onchain payment",
		})); err != nil {
			return err
		}
	}
	if err := w.Error(); err != nil {
		return fmt.Errorf("write reconciliation CSV: %w", err)
	}
	return nil
}

func safeCSVRow(row []string) []string {
	result := make([]string, len(row))
	for i, value := range row {
		trimmed := strings.TrimLeft(value, " \t\r\n")
		if strings.HasPrefix(trimmed, "=") || strings.HasPrefix(trimmed, "+") || strings.HasPrefix(trimmed, "-") || strings.HasPrefix(trimmed, "@") {
			value = "'" + value
		}
		result[i] = value
	}
	return result
}
